rustls/src/crypto/aws_lc_rs/tls13.rs

```diff
@@ -30,10 +30,12 @@ pub(crate) static TLS13_CHACHA20_POLY1305_SHA256_INTERNAL: &Tls13CipherSuite = &
     },
     hkdf_provider: &RingHkdf(hkdf::HKDF_SHA256, hmac::HMAC_SHA256),
     aead_alg: &Chacha20Poly1305Aead(AeadAlgorithm(&aead::CHACHA20_POLY1305)),
-    quic: Some(&super::quic::KeyBuilder(
-        &aead::CHACHA20_POLY1305,
-        &aead::quic::CHACHA20,
-    )),
+    quic: Some(&super::quic::KeyBuilder {
+        packet_alg: &aead::CHACHA20_POLY1305,
+        header_alg: &aead::quic::CHACHA20,
+        confidentiality_limit: u64::MAX,
+        integrity_limit: 1 << 36,
```

Comment:
Quinn has its own `PacketKey` trait with `confidentiality_limit()` and `integrity_limit()` methods. With rustls 0.22, the only realistic solution appears to be to implement this for `Box<dyn rustls::quic::PacketKey>`.

Reply:
So we store a `ring::quic::KeyBuilder` in the `Tls13CipherSuite` as the `&'static dyn quic::Algorithm`. The `KeyBuilder` is then used to generate `dyn PacketKey` values. Since the `KeyBuilder` is stored in the `Tls13CipherSuite`, I don't think it's easy to have it refer to things inside it? We can probably rework this for 0.23 but I'm not sure there's a good way to fix this without breaking 0.22 users.

Reply:
I've thought a bit more about this.

- `integrity_limit` in `CipherSuiteCommon` straight-out says it is meaningless for TCP-TLS, and exists only for the benefit of QUIC. If it is hard for quinn to use that, we can remove that from `CipherSuiteCommon` and just have it in the QUIC-specific code.
- `confidentiality_limit` in `CipherSuiteCommon` currently has the QUIC-specific values (based on 2**16 octet message sizes), which errs on the side of caution, but also a bit imprecise. If we have separate QUIC values, then we could change the documentation there to make it clear QUIC is not in scope when determining those values (ie `and 2 ** 16 for QUIC`)

Reply:
In case it's helpful [here's a branch](https://github.com/rustls/rustls/compare/main...cpu:rustls:cpu-quic-support-rebase) rebased on `main` that:

* Tries to integrate Ctz's thoughts from above:
  * `integrity_limit` is removed from `CipherSuiteCommon`
  * `confidentiality_limit` documentation is updated.
* Splits the change to start constructing quic keys from the crypto provider from the one introducing `AsRef` impls on `ClientConfig`/`ServerConfig` (related to [my comment](https://github.com/rustls/rustls/pull/1741#discussion_r1466515029)).

Reply:
Aside from queries on the `AsRef` thing I think I'd be happy to review this.

Reply:
:+1: I think the last barrier to get this branch moving is addressing a test and a fix for the append-hs deframer panic found upstream. I started on that, but haven't fixed it yet: https://github.com/cpu/rustls/pull/5